### PR TITLE
Cleanup our cache mess 

### LINKED ISSUE #74: redesign: MoveThreadToTop moves items out of view
When a message is received, the appropriate thread is moved, but the conversation list scrolls down so that the item in question is not displayed.

### LINKED ISSUE #84: simplify the outgoing messages cache
Back when the conversation view did not use full virtualization, we needed it to maintain references to the messages' UI representations.

Now we can easily get these by using a message's in-conversation index, and thus this might be the only thing we need to store in the cache.

### PR DESCRIPTION
**Pull request by Trolldemorted**
Our ID-based caches turned out to be messy, were problematic when implementing disappearing messages, and they were not contained in a single class. This commit simplifies things: The one and only storage of messages in the app's memory is now the VirtualizedMessagesCollection, which maintains DbIdToMessageMap so that we get a reference to a SignalMessage's Message.

closes #84

@golf1052 this should make supporting disappearing messages easier! Right now it breaks the visual representation of the unread marker and new identity messages, but I will work on that.

### CODE REVIEW

**Comment by Trolldemorted** [comment]
unread marker and new identity messages are fixed now!

File: Signal-Windows/Controls/VirtualizedMessagesCollection.cs [hunk 1]
```diff
@@ -196,18 +197,19 @@ public IEnumerator GetEnumerator()
 
         public int IndexOf(object value)
```
Comment (golf1052) [inline Signal-Windows/Controls/VirtualizedMessagesCollection.cs]: What is the purpose of this method? Why does it throw at the end?

**Comment by golf1052** [comment]
Looks good so far, I assume it's not complete yet?

File: Signal-Windows/Controls/VirtualizedMessagesCollection.cs (hunk 1 above)
Comment (Trolldemorted) [reply Signal-Windows/Controls/VirtualizedMessagesCollection.cs]: This method is called when you call ListView.ScrollIntoView. We only know the indices of the unread marker and of the last message (or we'd have to query every item), so we throw an exception if we are asked to return an index we don't know.

**Comment by Trolldemorted** [comment]
It should be complete, it works fine on my desktop and should make implementing disappearing messages much easier

**Comment by golf1052** [comment]
:shipit:

**Pull request merged**